[JENKINS-36629] Optimization for recent build stability calculation

stephenc opened the pull request:
[JENKINS-36629](https://issues.jenkins-ci.org/browse/JENKINS-36629)

Should at most load two builds into memory for the recently stable case. 

In the non-recently stable case we just pre-load the most recently stable one and skip loading any intermediary successful ones.

<table>
<tr>
<th>Probability of job failing</th><th>Expected number of jobs loaded</th>
</tr>
<tr><td>50%</td><td>4.56</td></tr>
<tr><td>40%</td><td>4.29</td></tr>
<tr><td>30%</td><td>3.92</td></tr>
<tr><td>20%</td><td>3.44</td></tr>
<tr><td>10%</td><td>2.80</td></tr>
<tr><td>5%</td><td>2.43</td></tr>
<tr><td>2%</td><td>2.17</td></tr>
<tr><td>1%</td><td>2.09</td></tr>
</table>


This compares with the previous implementation which would be expected to always load 5 jobs in the event that there were at least 5 jobs in the build history

@reviewbybees @jenkinsci/code-reviewers

stephenc committed:
Optimization for recent build stability calculation that should at most load two builds into memory for the recently stable case. In the non-recently stable case we just pre-load the most recently stable one and skip loading any intermediary successful ones

jglick commented:
This is related to [JENKINS-25078](https://issues.jenkins-ci.org/browse/JENKINS-25078) I suppose.

jglick commented:
Or [JENKINS-25075](https://issues.jenkins-ci.org/browse/JENKINS-25075) if you prefer.

jglick commented on core/src/main/java/hudson/model/Job.java:
```diff
@@ -1143,6 +1148,22 @@ private HealthReport getBuildStabilityHealthReport() {
         RunT i = getLastBuild();
+        RunT u = getLastFailedBuild();
+        if (i != null && u == null) {
+                // no failures, like ever
```
Or that were still present after `LogRotator` last ran.

stephenc replied:
it's only a metric... I am actually considering having it only consult the loaded builds ;-)

jglick replied:
I would be in favor of that. I am not sure `RunMap.getLoadedBuilds` as currently designed suffices, because it does not tell you which unmentioned build numbers in fact exist on disk—without actually trying to load them, you cannot tell whether they are there but unloaded, or have been deleted. There are a variety of reasons I think exposing `numberOnDisk` as an API in some form would be useful; this is one, as is a better fix of https://github.com/jenkinsci/jenkins/commit/0449883d152a772deea8621ce2e2c3e97e7e8a1f (JENKINS-23945).

oleg-nenashev replied:
All this logic is fragile in the case of concurrent builds, unfortunately. BTW, it's not a new defect

jglick commented on core/src/main/java/hudson/model/Job.java:
```diff
@@ -1143,6 +1148,22 @@ private HealthReport getBuildStabilityHealthReport() {
+            if (recentSuccessCount >= 5) {
+                return new HealthReport(100, Messages._Job_BuildStability(Messages._Job_NoRecentBuildFailed()));
+            } else {
+                // no need to load the intermediate builds, they're all successful, let's start iterating from
```
Are you sure calling `subMap` and then `size` does not force build loading? Offhand I would guess that it would. You need a test case using `RunLoadCounter` to prove that your fix is effective (and that the added complexity of this block did not introduce regressions—the correctness of this code is hardly self-evident).

jglick replied:
Perhaps [this code](https://github.com/jenkinsci/jenkins/blob/1c4c6432b795f807758152d06f3f8d9a7b9bf7bb/core/src/main/java/jenkins/model/lazy/BuildReferenceMapAdapter.java#L91) works, by avoiding `unwrap`. But needs to be mechanically verified.

stephenc replied:
Well inspection of the code says that it does not load the job until you call get. the sub-map is given the index and the size method only consults the index. (i.e. `core.size()`)

jglick replied:
Yes as I wrote above that _seems_ to be true, but this is pretty fragile. A test should be able to set up various combinations of recent builds with various statuses, and assert both the health result that is computed, and which builds are loaded to determine that.

jglick commented:
:bug: for complex new code without test coverage.

amuniz commented on core/src/main/java/hudson/model/Job.java:
```diff
@@ -61,15 +60,36 @@
 import hudson.widgets.HistoryWidget;
 import hudson.widgets.HistoryWidget.Adapter;
 import hudson.widgets.Widget;
+import java.awt.*;
```
If you are going to reorder imports it's a good chance to fix this.

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

stephenc committed:
[FIXED JENKINS-36629] Add a test

The test found out that RunMap's subMap implementation only returns loaded builds, so we needed to introduce a method to test if there was a specific run on the disk. For the weather reports we do not care if the record on the disk fails to load

stephenc commented:
@jglick happy now?

jglick commented on core/src/main/java/jenkins/model/lazy/AbstractLazyLoadRunMap.java:
```diff
@@ -292,6 +293,18 @@ public R get(int n) {
     }
 
     /**
+     * Checks if the disk index contains the specified build number.
```
I would suggest avoiding the term “disk” in APIs. It is kind of anachronistic, is it not? Perhaps just `buildExists`?

jglick commented on core/src/main/java/jenkins/model/lazy/AbstractLazyLoadRunMap.java:
```diff
@@ -292,6 +293,18 @@ public R get(int n) {
+     * @param number the build number to probe.
+     * @return {@code true} if there is an on disk entry for the corresponding number, note that this does not mean that
+     * the corresponding build record will load.
+     * @since 2.14
```
You are being optimistic…

jglick commented:
The test does help assure that the new code is not introducing regressions. It does nothing to demonstrate that the code is effective in blocking builds from being loaded. :bee: with reservations.

stephenc committed:
[JENKINS-36629] Use a name that does not assume on disk storage of run records

oleg-nenashev commented on core/src/main/java/hudson/model/Job.java:
```diff
@@ -1329,7 +1355,7 @@ public boolean equals(Object o) {
                     }
 
-                    public Color getColor() {
+                    public java.awt.Color getColor() {
```
full-qualified name is not required here

oleg-nenashev commented:
🐝 from me
